maingate/src/range.rs

```diff
@@ -215,19 +234,42 @@ impl<F: FieldExt> RangeChip<F> {
         composition_bit_lens: Vec<usize>,
         overflow_bit_lens: Vec<usize>,
     ) -> RangeConfig {
-        let mut overflow_bit_lens = overflow_bit_lens;
-        overflow_bit_lens.sort_unstable();
-        overflow_bit_lens.dedup();
-        let overflow_bit_lens: Vec<usize> =
-            overflow_bit_lens.into_iter().filter(|e| *e != 0).collect();
-
-        let mut composition_bit_lens = composition_bit_lens;
-        composition_bit_lens.sort_unstable();
-        composition_bit_lens.dedup();
-        let composition_bit_lens: Vec<usize> = composition_bit_lens
-            .into_iter()
-            .filter(|e| *e != 0)
-            .collect();
+        let [composition_bit_lens, overflow_bit_lens] = [composition_bit_lens, overflow_bit_lens]
+            .map(|mut bit_lens| {
+                bit_lens.sort_unstable();
+                bit_lens.dedup();
+                bit_lens
+            });
+
+        let has_overflow_limb = !overflow_bit_lens.is_empty();
+        let bit_len_tag = BTreeMap::from_iter(
+            BTreeSet::from_iter(composition_bit_lens.iter().chain(overflow_bit_lens.iter()))
+                .into_iter()
+                .enumerate()
+                .map(|(idx, bit_len)| (*bit_len, idx + 1)),
+        );
+
+        let t_tag = meta.lookup_table_column();
+        let t_value = meta.lookup_table_column();
+
+        macro_rules! lookup {
+            ($prefix:literal, $selector:ident, $tag:ident, $bit_lens:ident, [$($value:ident),*]) => {
+                $(
+                    meta.lookup(concat!($prefix, stringify!($value)), |meta| {
+                        let selector = meta.query_selector($selector);
+                        let tag = match $tag {
+                            Some(tag) =>  meta.query_fixed(tag, Rotation::cur()),
+                            None => {
+                                let tag = F::from(bit_len_tag[&$bit_lens[0]] as u64);
+                                selector.clone() * Expression::Constant(tag)
+                            },
+                        };
+                        let value = meta.query_advice($value, Rotation::cur());
+                        vec![(tag, t_tag), (selector * value, t_value)]
+                    });
+                )*
+            };
+        }
```

Comment:
Generally I try to avoid macros. But this one in particular is clear, easy to understand and with a really tiny scope. So it's fine at least for me to keep it as is.

Reply:
Anyway, feels like it could be replaced by a function or closure.

Reply:
Updated to function in https://github.com/privacy-scaling-explorations/halo2wrong/pull/38/commits/0ca396e35d51807a87de72e776a20081540808d4.